### DISCUSSION TITLE
WebRTC input failed

### DISCUSSION THREAD

**Discussion by martiankoder** [OvenMediaLabs/OvenMediaEngine#1173] [Help] [answered]
Hello everyone,
I have successfully streamed with RTMP using OBS. I am testing out WebRTC (OvenLiveKit) input but failed.
I followed the instructions from OvenLiveKit github here:
https://github.com/AirenSoft/OvenLiveKit-Web
I followed the step in the section Quick demo:
**Quick demo
step 1: You can see a quick demo in action by cloning the repository.
Clone repository
$ git clone https://github.com/AirenSoft/OvenLiveKit-Web.git
step 2: $ cd OvenLiveKit-Web
Install development dependencies.
$ npm install
step 3: Open the demo page using the WebPack's built-in web server.
$ npm run start**

After installing, then launch the index.html with google chrome. I gave the it the camera and mic permission. The video is on but I can't get the output stream.

I tried all possible ways with URL
1) using home computer IP address:
ws://192.168.0.10:3333/app/stream?direction=send&transport=tcp
wss://192.168.0.10:3333/app/stream?direction=send&transport=tcp
2) using my website subdomain (which I previously setup successfully streaming with RTMP input and WebRTC output):
wss://stream.mywebsite.com:3333/app/stream?direction=send&transport

I can see the video on index.html but actually it is not streaming. I inspect (right click -> inspect), I see the error:
`**OvenLiveKit.js:348 WebSocket connection to 'ws://192.168.0.19:3333/app/stream?direction=send&transport=tcp' failed:** `

'**webSocket.onerror Event {isTrusted: true, type: 'error', target: WebSocket, currentTarget: WebSocket, eventPhase: 2, …}isTrusted: truebubbles: falsecancelBubble: falsecancelable: falsecomposed: falsecurrentTarget: WebSocket {url: 'ws://192.168.0.19:3333/app/stream?direction=send&transport=tcp', readyState: 3, bufferedAmount: 0, onopen: ƒ, onerror: ƒ, …}defaultPrevented: falseeventPhase: 0returnValue: truesrcElement: WebSocket {url: 'ws://192.168.0.19:3333/app/stream?direction=send&transport=tcp', readyState: 3, bufferedAmount: 0, onopen: ƒ, onerror: ƒ, …}target: WebSocket {url: 'ws://192.168.0.19:3333/app/stream?direction=send&transport=tcp', readyState: 3, bufferedAmount: 0, onopen: ƒ, onerror: ƒ, …}timeStamp: 4646.200000047684type: "error"[[Prototype]]: Event**'

I got the same error when using the URL ( change in the index.html ):
wss://stream.mywebsite.com:3333/app/stream?direction=send&transport

I also check on https://demo.ovenplayer.com/ for input URL and output
Input WebRTC:
ws://192.168.0.10:3333/app/stream?direction=send&transport=tcp
wss://192.168.0.10:3333/app/stream?direction=send&transport=tcp
wss://stream.mywebsite.com:3333/app/stream?direction=send&transport
Output:
http://192.168.0.10:3333/app/stream?direction=whip&transport=tcp
https://192.168.0.10:3333/app/stream?direction=whip&transport=tcp
https://stream.mywebsite.com:3333/app/stream?direction=whip&transport

I tried all possible ways but can't see any output stream.

`Do you see what I am missing or any wrong? `

Any suggestion is greatly appreciated. Thank you

_### UPDATE 1:_ 

I tried to change the port from **3333** to **3334** , the input URL becomes
'wss://stream.mywebsite.ca:3334/app/stream?direction=send&transport=tcp'

1) when open index.html (right click on file), i got the error: (in inspect-> console on google chrome)
OvenLiveKit.js ==== Create WebRTC Input 1.1.0
OvenLiveKit.js:210 OvenLiveKit.js : Request Stream To Input Devices With Constraints Object
OvenLiveKit.js:215 OvenLiveKit.js : Received Media Stream From Input Device MediaStream
OvenLiveKit.js:785 OvenLiveKit.js ==== Start Streaming with connectionConfig undefined
OvenLiveKit.js:370 webSocket.onmessage Cannot create offer
s.onmessage @ OvenLiveKit.js:370

2) if run with **npm run start** in Visual studio code, i got error:(in inspect-> console on google chrome)
OvenLiveKit.js ==== Create WebRTC Input 1.1.0
OvenLiveKit.js:210 OvenLiveKit.js : Request Stream To Input Devices With Constraints Object
OvenLiveKit.js:215 OvenLiveKit.js : Received Media Stream From Input Device MediaStream
OvenLiveKit.js:785 OvenLiveKit.js ==== Start Streaming with connectionConfig undefined
:8080/favicon.ico:1          Failed to load resource: the server responded with a status of 404 (Not Found)
OvenLiveKit.js:370 webSocket.onmessage Cannot create offer

``
Ran the second time, the error is different (I said receive stream but, still has errors):

`OvenLiveKit.js : Received Media Stream From Input Device MediaStreamactive: trueid: "42325d48-a3f2-4b16-b97e-1cda7276d0e0"onactive: nullonaddtrack: nulloninactive: nullonremovetrack: null[[Prototype]]: MediaStreamactive: (...)addTrack: ƒ addTrack()clone: ƒ clone()getAudioTracks: ƒ getAudioTracks()getTrackById: ƒ getTrackById()getTracks: ƒ getTracks()getVideoTracks: ƒ getVideoTracks()id: (...)onactive: (...)onaddtrack: (...)oninactive: (...)onremovetrack: (...)removeTrack: ƒ removeTrack()constructor: ƒ MediaStream()Symbol(Symbol.toStringTag): "MediaStream"get active: ƒ active()get id: ƒ id()get onactive: ƒ onactive()set onactive: ƒ onactive()get onaddtrack: ƒ onaddtrack()set onaddtrack: ƒ onaddtrack()get oninactive: ƒ oninactive()set oninactive: ƒ oninactive()get onremovetrack: ƒ onremovetrack()set onremovetrack: ƒ onremovetrack()[[Prototype]]: EventTarget
OvenLiveKit.js:785 OvenLiveKit.js ==== Start Streaming with connectionConfig undefined
:8080/favicon.ico:1          Failed to load resource: the server responded with a status of 404 (Not Found)
OvenLiveKit.js:581 OvenLiveKit.js : Create Peer Connection With Config Object
OvenLiveKit.js:590 OvenLiveKit.js : Add Track To Peer Connection MediaStreamTrack
OvenLiveKit.js:590 OvenLiveKit.js : Add Track To Peer Connection MediaStreamTrack
OvenLiveKit.js:611 OvenLiveKit.js : Modified offer sdp

v=0
o=OvenMediaEngine 534 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE fsqumw zV0LBk
a=msid-semantic:WMS jbenLm6OWUB9rVK0shfA1dvlMQ3pJw8qZ4aG
a=fingerprint:sha-256 A7:80:82:36:8F:93:7C:05:E8:C3:D4:20:44:2E:34:9F:19:B6:4C:F8:CB:78:66:ED:46:6A:C3:4A:37:CF:A0:A0
a=ice-options:trickle
a=ice-ufrag:pUirhd
a=ice-pwd:ZNU8BcmqHM16uD3OEAjkSv4eW9Xirs2Q
m=video 9 UDP/TLS/RTP/SAVPF 100 101
c=IN IP4 0.0.0.0
a=recvonly
a=mid:fsqumw
a=setup:actpass
a=rtcp-mux
a=msid:jbenLm6OWUB9rVK0shfA1dvlMQ3pJw8qZ4aG KVa5DGUFHbnMXfl9E0sOLNti7SrJodyqcQ8A
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=rtpmap:100 H264/90000
a=fmtp:100 packetization-mode=1;profile-level-id=42e01f;level-asymmetry-allowed=1
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack pli
a=rtpmap:101 VP8/90000
a=rtcp-fb:101 transport-cc
a=rtcp-fb:101 ccm fir
a=rtcp-fb:101 nack pli
m=audio 9 UDP/TLS/RTP/SAVPF 102
c=IN IP4 0.0.0.0
a=recvonly
a=mid:zV0LBk
a=setup:actpass
a=rtcp-mux
a=msid:jbenLm6OWUB9rVK0shfA1dvlMQ3pJw8qZ4aG 1isSqwjpr2tfP7IURkn3a56xWcXvmO8uMV4G
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=rtpmap:102 OPUS/48000/2
a=fmtp:102 sprop-stereo=1;stereo=1;minptime=10;useinbandfec=1
a=rtcp-fb:102 transport-cc

OvenLiveKit.js:628 OvenLiveKit.js : Modified answer sdp

v=0
o=- 4381331846729771319 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE fsqumw zV0LBk
a=msid-semantic: WMS 42325d48-a3f2-4b16-b97e-1cda7276d0e0
m=video 9 UDP/TLS/RTP/SAVPF 100 101
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Yff0
a=ice-pwd:CPOqAtLHk6wVPoqQWB+6c9NH
a=ice-options:trickle
a=fingerprint:sha-256 A9:06:4D:BD:1E:E1:4B:86:B4:C3:06:CB:2C:CF:36:81:5F:5C:2F:04:49:58:02:4A:37:7F:B4:15:23:FE:69:38
a=setup:active
a=mid:fsqumw
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=sendonly
a=msid:42325d48-a3f2-4b16-b97e-1cda7276d0e0 d6f7057c-3d0e-4245-b5e9-5cc2835f424a
a=rtcp-mux
a=rtpmap:100 H264/90000
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack pli
a=fmtp:100 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:101 VP8/90000
a=rtcp-fb:101 transport-cc
a=rtcp-fb:101 ccm fir
a=rtcp-fb:101 nack pli
a=ssrc:4246789658 cname:Paht4gewLoSP/km3
m=audio 9 UDP/TLS/RTP/SAVPF 102
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:Yff0
a=ice-pwd:CPOqAtLHk6wVPoqQWB+6c9NH
a=ice-options:trickle
a=fingerprint:sha-256 A9:06:4D:BD:1E:E1:4B:86:B4:C3:06:CB:2C:CF:36:81:5F:5C:2F:04:49:58:02:4A:37:7F:B4:15:23:FE:69:38
a=setup:active
a=mid:zV0LBk
a=extmap:3 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=sendonly
a=msid:42325d48-a3f2-4b16-b97e-1cda7276d0e0 5aed60c9-6f6f-49b2-af3c-0842a0aafe52
a=rtcp-mux
a=rtpmap:102 OPUS/48000/2
a=rtcp-fb:102 transport-cc
a=fmtp:102 minptime=10;useinbandfec=1
a=ssrc:1654251786 cname:Paht4gewLoSP/km3`

**Comment by getroot** [comment, marked as answer]
WebRTC device access can only be done from a Secured URL. That is, it must be an https site where the certificate is installed, such as https://demo.ovenplayer.com/demo_input.html .

At this time, since the browser does not allow mixed content, OME must use the wss URL.

Your wss url is the same as ws url with port 3333. TLSPort is probably 3334 if you are using the default settings. Did you install the certificate?

When posting a question, uploading your Server.xml and ovenmediaengine.log files will save our energy.

**Comment by martiankoder** [comment]
Thank you for your quick reply. Thank you so much. It's running now.
